File: app/python/src/t7oolkit/utils/image_resize.py
from __future__ import annotations

from concurrent.futures import ThreadPoolExecutor, as_completed
from dataclasses import dataclass
from pathlib import Path

from PIL import Image

IMAGE_EXTENSIONS = {".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp", ".tiff"}


@dataclass(frozen=True)
class ResizeResult:
    filename: str
    source_size: tuple[int, int]
    output_size: tuple[int, int]
    skipped: bool = False


@dataclass(frozen=True)
class ResizeError:
    filename: str
    message: str


def list_image_files(input_dir: str | Path) -> list[Path]:
    directory = Path(input_dir)
    return sorted(
        path
        for path in directory.iterdir()
        if path.is_file() and path.suffix.lower() in IMAGE_EXTENSIONS
    )


def resize_image(input_path: str | Path, output_path: str | Path, max_size: int = 640) -> ResizeResult:
    """Resize image to fit within max_size x max_size while preserving aspect ratio."""
    input_path = Path(input_path)
    output_path = Path(output_path)
    output_path.parent.mkdir(parents=True, exist_ok=True)

    with Image.open(input_path) as img:
        source_size = (img.width, img.height)
        if img.width <= max_size and img.height <= max_size:
            img.save(output_path)
            return ResizeResult(
                filename=input_path.name,
                source_size=source_size,
                output_size=source_size,
                skipped=True,
            )

        ratio = min(max_size / img.width, max_size / img.height)
        new_size = (int(img.width * ratio), int(img.height * ratio))
        resized = img.resize(new_size, Image.Resampling.LANCZOS)
        resized.save(output_path)
        return ResizeResult(
            filename=input_path.name,
            source_size=source_size,
            output_size=new_size,
        )


def batch_resize_images(
    input_dir: str | Path,
    output_dir: str | Path,
    *,
    max_size: int = 640,
    workers: int = 4,
) -> tuple[list[ResizeResult], list[ResizeError]]:
    input_dir = Path(input_dir)
    output_dir = Path(output_dir)
    output_dir.mkdir(parents=True, exist_ok=True)

    files = list_image_files(input_dir)
    if not files:
        return [], []

    results: list[ResizeResult] = []
    errors: list[ResizeError] = []

    def process(path: Path) -> ResizeResult | ResizeError:
        try:
            return resize_image(path, output_dir / path.name, max_size=max_size)
        except Exception as exc:
            return ResizeError(filename=path.name, message=str(exc))

    worker_count = max(1, min(workers, len(files)))
    with ThreadPoolExecutor(max_workers=worker_count) as executor:
        futures = {executor.submit(process, path): path for path in files}
        for future in as_completed(futures):
            outcome = future.result()
            if isinstance(outcome, ResizeError):
                errors.append(outcome)
            else:
                results.append(outcome)

    results.sort(key=lambda item: item.filename.lower())
    errors.sort(key=lambda item: item.filename.lower())
    return results, errors


def format_result_line(result: ResizeResult) -> str:
    if result.skipped:
        return f"{result.filename}: {result.source_size[0]}x{result.source_size[1]} (无需缩放)"
    return (
        f"{result.filename}: {result.source_size[0]}x{result.source_size[1]} "
        f"-> {result.output_size[0]}x{result.output_size[1]}"
    )

File: app/tauri/src-tauri/src/image_resize.rs
use image::imageops::FilterType;
use image::GenericImageView;
use rayon::prelude::*;
use serde::Serialize;
use std::fs;
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Emitter};

const IMAGE_EXTENSIONS: &[&str] = &[
    ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp", ".tiff",
];

#[derive(Debug, Clone, Serialize)]
pub struct ResizeResult {
    pub filename: String,
    pub source_size: (u32, u32),
    pub output_size: (u32, u32),
    pub skipped: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct ResizeError {
    pub filename: String,
    pub message: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct BatchResizeResponse {
    pub results: Vec<ResizeResult>,
    pub errors: Vec<ResizeError>,
}

#[derive(Debug, Clone, Serialize)]
pub struct ResizeProgressPayload {
    pub current: usize,
    pub total: usize,
    pub filename: String,
}

pub fn list_image_files(input_dir: &Path) -> Vec<PathBuf> {
    let Ok(entries) = fs::read_dir(input_dir) else {
        return Vec::new();
    };

    let mut files: Vec<PathBuf> = entries
        .filter_map(Result::ok)
        .map(|entry| entry.path())
        .filter(|path| path.is_file())
        .filter(|path| {
            path.extension()
                .and_then(|ext| ext.to_str())
                .map(|ext| {
                    let lower = format!(".{}", ext.to_lowercase());
                    IMAGE_EXTENSIONS.contains(&lower.as_str())
                })
                .unwrap_or(false)
        })
        .collect();

    files.sort_by_key(|path| path.file_name().map(|name| name.to_ascii_lowercase()));
    files
}

fn resize_image(
    input_path: &Path,
    output_path: &Path,
    max_size: u32,
) -> Result<ResizeResult, String> {
    if let Some(parent) = output_path.parent() {
        fs::create_dir_all(parent).map_err(|err| err.to_string())?;
    }

    let img = image::open(input_path).map_err(|err| err.to_string())?;
    let (width, height) = img.dimensions();
    let source_size = (width, height);

    if width <= max_size && height <= max_size {
        img.save(output_path).map_err(|err| err.to_string())?;
        return Ok(ResizeResult {
            filename: input_path
                .file_name()
                .and_then(|name| name.to_str())
                .unwrap_or("")
                .to_string(),
            source_size,
            output_size: source_size,
            skipped: true,
        });
    }

    let ratio = (max_size as f64 / width as f64).min(max_size as f64 / height as f64);
    let new_width = (width as f64 * ratio).round() as u32;
    let new_height = (height as f64 * ratio).round() as u32;
    let resized = img.resize(new_width, new_height, FilterType::Lanczos3);
    resized.save(output_path).map_err(|err| err.to_string())?;

    Ok(ResizeResult {
        filename: input_path
            .file_name()
            .and_then(|name| name.to_str())
            .unwrap_or("")
            .to_string(),
        source_size,
        output_size: (new_width, new_height),
        skipped: false,
    })
}

pub fn batch_resize_images(
    input_dir: &Path,
    output_dir: &Path,
    max_size: u32,
    workers: usize,
    app: &AppHandle,
) -> BatchResizeResponse {
    fs::create_dir_all(output_dir).ok();

    let files = list_image_files(input_dir);
    if files.is_empty() {
        return BatchResizeResponse {
            results: Vec::new(),
            errors: Vec::new(),
        };
    }

    let pool = rayon::ThreadPoolBuilder::new()
        .num_threads(workers.max(1).min(files.len()))
        .build()
        .unwrap_or_else(|_| rayon::ThreadPoolBuilder::new().build().expect("rayon pool"));

    let total = files.len();
    let progress = std::sync::atomic::AtomicUsize::new(0);

    let outcomes: Vec<Result<ResizeResult, ResizeError>> = pool.install(|| {
        files
            .par_iter()
            .map(|path| {
                let current = progress.fetch_add(1, std::sync::atomic::Ordering::SeqCst) + 1;
                let filename = path
                    .file_name()
                    .and_then(|name| name.to_str())
                    .unwrap_or("")
                    .to_string();

                let _ = app.emit(
                    "resize-progress",
                    ResizeProgressPayload {
                        current,
                        total,
                        filename: filename.clone(),
                    },
                );

                match resize_image(path, &output_dir.join(path.file_name().unwrap_or_default()), max_size) {
                    Ok(result) => Ok(result),
                    Err(message) => Err(ResizeError { filename, message }),
                }
            })
            .collect()
    });

    let mut results = Vec::new();
    let mut errors = Vec::new();

    for outcome in outcomes {
        match outcome {
            Ok(result) => results.push(result),
            Err(error) => errors.push(error),
        }
    }

    results.sort_by_key(|item| item.filename.to_ascii_lowercase());
    errors.sort_by_key(|item| item.filename.to_ascii_lowercase());

    BatchResizeResponse { results, errors }
}

#[tauri::command]
pub fn batch_resize_images_command(
    app: AppHandle,
    input_dir: String,
    output_dir: String,
    max_size: u32,
    workers: u32,
) -> BatchResizeResponse {
    batch_resize_images(
        Path::new(&input_dir),
        Path::new(&output_dir),
        max_size,
        workers as usize,
        &app,
    )
}
